packages/runtime/container-runtime/src/channelCollection.ts

```diff
@@ -1148,6 +1152,36 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
 		return builder.getSummaryTree();
 	}
 
+	/**
+	 * Gets the GC data. Used when attaching or serializing a detached container.
+	 */
+	public getAttachGCData(telemetryContext?: ITelemetryContext): IGarbageCollectionData {
+		const builder = new GCDataBuilder();
+		// Attaching graph of some stores can cause other stores to get bound too.
+		// So keep taking summary until no new stores get bound.
+		let notBoundContextsLength: number;
+		do {
+			notBoundContextsLength = this.contexts.notBoundLength();
+			// Iterate over each data store and ask for its GC data.
+			Array.from(this.contexts)
+				.filter(
+					([key, _]) =>
+						// Take GC data of bounded data stores only.
+						!this.contexts.isNotBound(key),
```

Comment:
tl;dr - I would strongly encourage you to use the [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) (similar [example](https://github.com/microsoft/FluidFramework/blob/c6293d4b833da067c9abbd447eb9483fc8425be9/packages/runtime/datastore/src/dataStoreRuntime.ts#L924)) to share the structure of this do-while loop between both functions here.

Reasons/Thoughts:
* The logic should be the same, but people won't necessarily know to update both places
* To that point... you are already diverging the logic on this line.  Shouldn't it check the same 3 conditions as above?
* Are there other places the logic differs, and can we keep it the same so we can share the code?

Reply:
I made a change for this but it looks slightly risky. Let me do this in a follow-up PR.